src/shared/particles/base_particles.cpp

```diff
@@ -365,6 +367,61 @@ void BaseParticles::readFromXmlForReloadParticle(std::string &filefullpath)
     DataAssembleOperation<loopParticleVariables> loop_variable_namelist;
     loop_variable_namelist(all_particle_data_, variables_to_reload_, read_variable_from_xml);
 }
+
+    void BaseParticles::registerDeviceMemory() {
+        registerDeviceVariable<DeviceVecd>("Position", pos_.size(), pos_.data());
```

Comment:
You might want to use `total_real_particles_` instead on relying on vector size to avoid copy unnecessary buffer particles (unless particles switching and removal should also be performed on device)

Reply:
You are right, I can use `total_real_particles_`. I am wondering if those cases in which `total_real_particles_ != pos_.size()` would mean that `registerDeviceMemory` have been called at the wrong time, and if it should be handled with an error or a special behavior. Or it could also be considered as a precondition of `registerDeviceMemory` to have the right amount of particles.

Reply:
The main case when `total_real_particles_ != pos_.size()` is for fluid when there are inlet/outlet which require buffer particles (I don't know how the notion of ghost particle is used within SPHinXsys). Depending on the size of the domain this can make quite a difference.